src/Http/Routing/src/RoutingMetrics.cs

```diff
@@ -0,0 +1,43 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.Metrics;
+using Microsoft.Extensions.Diagnostics.Metrics;
+
+namespace Microsoft.AspNetCore.Routing;
+
+internal sealed class RoutingMetrics
+{
+    public const string MeterName = "Microsoft.AspNetCore.Routing";
+
+    private readonly Meter _meter;
+    private readonly Counter<long> _matchSuccessCounter;
+    private readonly Counter<long> _matchFailureCounter;
+
+    public RoutingMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.Create(MeterName);
+
+        _matchSuccessCounter = _meter.CreateCounter<long>(
+           "routing-match-success",
+            description: "Number of requests successfully matched to an endpoint by routing.");
+
+        _matchFailureCounter = _meter.CreateCounter<long>(
+           "routing-match-failure",
+            description: "Number of requests that failed to match to an endpoint by routing.");
+    }
+
+    public bool MatchSuccessCounterEnabled => _matchSuccessCounter.Enabled;
+
+    public void MatchSuccess(string route, bool isFallback)
+    {
+        _matchSuccessCounter.Add(1,
+            new KeyValuePair<string, object?>("route", route),
+            new KeyValuePair<string, object?>("fallback", isFallback));
+    }
+
+    public void MatchFailure()
+    {
+        _matchFailureCounter.Add(1);
```

Comment:
We'll need to be clear in the docs that there are lots of requests handled by middleware that will be reported as route match failures. E.g. static files, auth callbacks, etc. I wonder if that will make this counter more noise than value.

Reply:
@tekian ?

Reply:
@geeknoid @tekian Is this a counter you want?

Reply:
> We'll need to be clear in the docs that there are lots of requests handled by middleware that will be reported as route match failures. E.g. static files, auth callbacks, etc. I wonder if that will make this counter more noise than value.

Can you show an example of this? I'm not sure how desirable this is. We need to verify this for all sorts of applications. (blazor to apis)

Reply:
I thought about this counter more, and I think we should keep it.
* It doesn't cost anything to have it. Implementation is simple, and the counter doesn't hurt perf if no one listens. If someone doesn't care about it, then they won't listen.
* If people see a counter that reports route matching success, they'll expect a counter that reports failure. Provides symmetry.
* Potential confusion about what the counter reports and what it means can be explained in the description. The description can say that unmatched requests may be handled later by other middleware, such as static files or authentication callbacks.

I'll make these changes and then merge this PR soon. If I don't hear feedback, then I'll assume people are ok with this. It's not an end to discussion, I just want to hold up metrics progress towards done.